From 440e45d14b7d86e3467024a2ec2c3750c1d9f3ca Mon Sep 17 00:00:00 2001
From: Alex Burmashev <alexander.burmashev@oracle.com>
Date: Tue, 11 Nov 2025 12:03:18 +0000
Subject: [PATCH] OpenELA 9.6 debranding update

---
 SPECS/gnome-settings-daemon.spec | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/SPECS/gnome-settings-daemon.spec b/SPECS/gnome-settings-daemon.spec
index 9033846..4659d87 100644
--- a/SPECS/gnome-settings-daemon.spec
+++ b/SPECS/gnome-settings-daemon.spec
@@ -9,6 +9,10 @@

 %global tarball_version %%(echo %{version} | tr '~' '.')

+%if %{undefined openela}
+%bcond_without subman
+%endif
+
 Name:           gnome-settings-daemon
 Version:        40.0.1
 Release:        22%{?dist}
@@ -36,7 +40,9 @@ BuildRequires:  pkgconfig(gsettings-desktop-schemas) >= %{gsettings_desktop_sche
 BuildRequires:  pkgconfig(gtk+-3.0) >= %{gtk3_version}
 BuildRequires:  pkgconfig(gudev-1.0)
 BuildRequires:  pkgconfig(gweather-3.0) >= %{libgweather_version}
+%if %{with subman}
 BuildRequires:  pkgconfig(json-glib-1.0)
+%endif
 BuildRequires:  pkgconfig(lcms2) >= 2.2
 BuildRequires:  pkgconfig(libcanberra-gtk3)
 BuildRequires:  pkgconfig(libgeoclue-2.0)
@@ -68,7 +74,9 @@ Requires: gtk3%{?_isa} >= %{gtk3_version}
 Requires: libgweather%{?_isa} >= %{libgweather_version}

 Patch00001: 0001-power-Enable-power-saver-profile-when-low-on-battery.patch
+%if %{with subman}
 Patch00002: subscription-manager-support.patch
+%endif
 Patch00003: im-module-setting-fix.patch
 Patch00004: print-notifications-clear-in-stop.patch
 Patch00005: 0001-power-Respect-the-nothing-power-button-action-for-VM.patch
@@ -186,11 +194,13 @@ cp %{SOURCE1} $RPM_BUILD_ROOT%{_datadir}/glib-2.0/schemas
 %{_libexecdir}/gsd-wwan
 %{_datadir}/glib-2.0/schemas/org.gnome.settings-daemon.plugins.wwan.gschema.xml

+%if %{with subman}
 %{_sysconfdir}/xdg/autostart/org.gnome.SettingsDaemon.Subscription.desktop
 %{_libexecdir}/gsd-subman
 %{_libexecdir}/gsd-subman-helper
 %{_datadir}/polkit-1/actions/org.gnome.settings-daemon.plugins.subman.policy
 %{_datadir}/polkit-1/rules.d/org.gnome.settings-daemon.plugins.subman.rules
+%endif

 %dir %{_libdir}/gnome-settings-daemon-40
 %{_libdir}/gnome-settings-daemon-40/libgsd.so
--
2.47.3
