diff --git a/SPECS/gnome-settings-daemon.spec b/SPECS/gnome-settings-daemon.spec
index a559b4a..2f45dce 100644
--- a/SPECS/gnome-settings-daemon.spec
+++ b/SPECS/gnome-settings-daemon.spec
@@ -99,6 +99,7 @@ Patch30001: 0001-media-keys-Mark-screen-cast-as-stopped-if-it-was-sig.patch
 Patch40001: 0001-housekeeping-Add-a-GPU-memory-usage-notification.patch
 
 # subscription manager integration
+%if %{with subman}
 Patch50001: 0001-subman-Add-a-new-plugin-to-provide-system-subscripti.patch
 Patch50002: 0002-subman-Add-InstalledProducts-dbus-property-for-g-c-c.patch
 Patch50003: 0003-subman-Increase-RHSM-dbus-call-timeouts.patch
@@ -115,6 +116,7 @@ Patch50013: 0013-subman-Improve-subscription-status-handling.patch
 Patch50014: 0014-subman-Drop-LAST-from-status-enum.patch
 Patch50015: 0015-subman-Clean-up-notification-behavior.patch
 Patch50016: 0016-subman-Update-POTFILES.in.patch
+%endif
 # https://bugzilla.redhat.com/show_bug.cgi?id=1876291
 Patch50017: 0017-Update-translations.patch
 
@@ -237,11 +239,13 @@ mkdir $RPM_BUILD_ROOT%{_libdir}/gnome-settings-daemon-3.0/gtk-modules
 %{_sysconfdir}/xdg/autostart/org.gnome.SettingsDaemon.Sharing.desktop
 %{_datadir}/glib-2.0/schemas/org.gnome.settings-daemon.plugins.sharing.gschema.xml
 
+%if %{with subman}
 %{_sysconfdir}/xdg/autostart/org.gnome.SettingsDaemon.Subscription.desktop
 %{_libexecdir}/gsd-subman
 %{_libexecdir}/gsd-subman-helper
 %{_datadir}/polkit-1/actions/org.gnome.settings-daemon.plugins.subman.policy
 %{_datadir}/polkit-1/rules.d/org.gnome.settings-daemon.plugins.subman.rules
+%endif
 
 %{_libdir}/gnome-settings-daemon-3.0/libgsd.so
 
