From d38a7f2913b30c5c3954f32cf54068c85cb758b0 Mon Sep 17 00:00:00 2001
From: Kalev Lember <klember@redhat.com>
Date: Fri, 12 Feb 2021 14:51:29 +0100
Subject: [PATCH 16/16] subman: Update POTFILES.in

---
 po/POTFILES.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/po/POTFILES.in b/po/POTFILES.in
index 7f17d5b5..9b94dc0f 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -1,30 +1,32 @@
 # Files with translatable strings.
 # Please keep this file in alphabetical order.
 data/org.gnome.settings-daemon.peripherals.gschema.xml.in
 data/org.gnome.settings-daemon.peripherals.wacom.gschema.xml.in
 data/org.gnome.settings-daemon.plugins.color.gschema.xml.in
 data/org.gnome.settings-daemon.plugins.gschema.xml.in
 data/org.gnome.settings-daemon.plugins.housekeeping.gschema.xml.in
 data/org.gnome.settings-daemon.plugins.media-keys.gschema.xml.in
 data/org.gnome.settings-daemon.plugins.power.gschema.xml.in
 data/org.gnome.settings-daemon.plugins.sharing.gschema.xml.in
 data/org.gnome.settings-daemon.plugins.xsettings.gschema.xml.in
 plugins/color/gsd-color-calibrate.c
 plugins/color/gsd-color-manager.c
 plugins/datetime/gsd-datetime-manager.c
 plugins/housekeeping/gsd-disk-space.c
 plugins/keyboard/gsd-keyboard-manager.c
 plugins/media-keys/gsd-media-keys-manager.c
 plugins/media-keys/gsd-screenshot-utils.c
 plugins/media-keys/shortcuts-list.h
 plugins/mouse/gsd-mouse-manager.c
 plugins/power/gpm-common.c
 plugins/power/gsd-power-manager.c
 plugins/power/org.gnome.settings-daemon.plugins.power.policy.in.in
 plugins/print-notifications/gsd-printer.c
 plugins/print-notifications/gsd-print-notifications-manager.c
 plugins/smartcard/gsd-smartcard-manager.c
 plugins/smartcard/gsd-smartcard-service.c
+plugins/subman/gsd-subscription-manager.c
+plugins/subman/org.gnome.settings-daemon.plugins.subman.policy.in.in
 plugins/wacom/gsd-wacom-manager.c
 plugins/wacom/org.gnome.settings-daemon.plugins.wacom.policy.in.in
 plugins/xsettings/gsd-xsettings-manager.c
-- 
2.41.0.rc2

